src/llvm-expand-atomic-modify.cpp

```diff
@@ -0,0 +1,468 @@
+          return false;
+        Ret = cast<ReturnInst>(BB.getTerminator());
+      }
+    }
+    if (Ret == nullptr)
+      return false;
+    // Now examine the instruction list
+    RetVal = Ret->getReturnValue();
+    if (!RetVal->hasOneUse())
+      return false;
+  }
+  if (RetVal == Old) {
+    // special token indicating to convert to an atomic fence
+    if (ValOp) *ValOp = nullptr;
+    return AtomicRMWInst::Or;
+  }
+  if (Old->use_empty()) {
+    if (ValOp) *ValOp = nullptr;
+    return AtomicRMWInst::Xchg;
+  }
+  if (auto BinOp = dyn_cast<BinaryOperator>(RetVal)) {
+    if ((BinOp->getOperand(0) == Old || (BinOp->isCommutative() && BinOp->getOperand(1) == Old)) && canReorderWithRMW(*BinOp, verifyop)) {
+      if (ValOp) *ValOp = &BinOp->getOperandUse(BinOp->getOperand(0) == Old ? 1 : 0);
+      switch (BinOp->getOpcode()) {
+        case Instruction::Add:
+          return AtomicRMWInst::Add;
+        case Instruction::Sub:
+          return AtomicRMWInst::Sub;
+        case Instruction::And:
+          return AtomicRMWInst::And;
+        case Instruction::Or:
+          return AtomicRMWInst::Or;
+        case Instruction::Xor:
+          return AtomicRMWInst::Xor;
+        case Instruction::FAdd:
+          return AtomicRMWInst::FAdd;
+        case Instruction::FSub:
+          return AtomicRMWInst::FSub;
+        default:
+          break;
+      }
+    }
+    if (BinOp->getOpcode() == Instruction::Xor) {
+      if (auto CI = dyn_cast<ConstantInt>(BinOp->getOperand(1))) {
+        if (CI->isAllOnesValue()) {
+          BinOp = dyn_cast<BinaryOperator>(BinOp->getOperand(0));
+          if (BinOp && BinOp->hasOneUse() && BinOp->getOpcode() == Instruction::And) {
+            if ((BinOp->getOperand(0) == Old || (BinOp->isCommutative() && BinOp->getOperand(1) == Old)) && canReorderWithRMW(*BinOp, verifyop)) {
+              if (ValOp) *ValOp = &BinOp->getOperandUse(BinOp->getOperand(0) == Old ? 1 : 0);
+              return AtomicRMWInst::Nand;
+            }
+          }
+        }
+      }
+    }
+    return false;
+  } else if (auto Intr = dyn_cast<IntrinsicInst>(RetVal)) {
+    if (Intr->arg_size() == 2) {
+      if ((Intr->getOperand(0) == Old || (Intr->isCommutative() && Intr->getOperand(1) == Old)) && canReorderWithRMW(*Intr, verifyop)) {
+        if (ValOp) *ValOp = &Intr->getOperandUse(Intr->getOperand(0) == Old ? 1 : 0);
+        switch (Intr->getIntrinsicID()) {
+          case Intrinsic::minnum:
+            return AtomicRMWInst::FMin;
+          case Intrinsic::maxnum:
+            return AtomicRMWInst::FMax;
+          case Intrinsic::smax:
+            return AtomicRMWInst::Max;
+          case Intrinsic::umax:
+            return AtomicRMWInst::UMax;
+          case Intrinsic::smin:
+            return AtomicRMWInst::Min;
+          case Intrinsic::umin:
+            return AtomicRMWInst::UMin;
+          // for LLVM 19:
+          // case Intrinsic::usub_sat:
+          //  return AtomicRMWInst::USubSat;
+        }
```

Comment:
Should we already turn this into an ifdef?

Reply:
I couldn't test, so I would rather have slightly less optimal runtime than an error that I couldn't test beforehand